query-service-impl/src/test/java/org/hypertrace/core/query/service/postgres/QueryRequestToPostgresSQLConverterTest.java

```diff
@@ -1150,13 +1151,143 @@ void testQueryWithAverageRateInOrderBy() {
             + TENANT_ID
             + "' "
             + "and ( start_time_millis >= 1637297304041 and start_time_millis < 1637300904041 and service_id != 'null' ) "
-            + "group by service_id, service_name "
+            + "group by 1, 2 "
             + "order by SUM(error_count) / 3600.0 "
             + "limit 10000",
         tableDefinition,
         executionContext);
   }
 
+  @Test
+  void testQueryWithDistinctCountAggregationAndGroupByForArrayColumn() {
+    Filter startTimeFilter =
+        createTimeFilter("Span.start_time_millis", Operator.GT, 1570658506605L);
+    Filter endTimeFilter = createTimeFilter("Span.end_time_millis", Operator.LT, 1570744906673L);
+    QueryRequest queryRequest =
+        QueryRequest.newBuilder()
+            .addSelection(createColumnExpression("Span.id"))
+            .addGroupBy(createColumnExpression("Span.id"))
+            .addAggregation(
+                createAliasedFunctionExpression(
+                    "DISTINCTCOUNT", "Span.labels", "distinctcount_labels"))
+            .addAggregation(
+                createAliasedFunctionExpression(
+                    "AVG", "Span.duration_millis", "avg_duration_millis"))
+            .addAggregation(
+                createAliasedFunctionExpression(
+                    "MAX", "Span.duration_millis", "avg_duration_millis"))
+            .setFilter(
+                Filter.newBuilder()
+                    .setOperator(Operator.AND)
+                    .addChildFilter(startTimeFilter)
+                    .addChildFilter(endTimeFilter)
+                    .build())
+            .addOrderBy(
+                createOrderByExpression(
+                    createAliasedFunctionExpression(
+                        "DISTINCTCOUNT", "Span.labels", "distinctcount_labels"),
+                    SortOrder.ASC))
+            .setLimit(15)
+            .build();
+
+    TableDefinition tableDefinition = getDefaultTableDefinition();
+    defaultMockingForExecutionContext();
+
+    assertSQLQuery(
+        queryRequest,
+        "select encode(span_id, 'hex'), count(distinct column2),"
+            + " avg(duration_millis), max(duration_millis) FROM"
+            + " ( select span_id, unnest(labels) as column2, duration_millis from public.\"span-event-view\""
```

Comment:
So the logic is to fetch all the columns required in the main query?

Reply:
That is correct